src/strands/multiagent/swarm.py

```diff
@@ -0,0 +1,646 @@
+    max_iterations: int = 20
+    execution_timeout: float = 900.0  # Total execution timeout (seconds)
+    node_timeout: float = 300.0  # Individual node timeout (seconds)
+    ping_pong_check_nodes: int = 8  # Number of recent nodes to check for ping-pong
+    ping_pong_min_unique_nodes: int = 3  # Minimum unique nodes required in recent sequence
+
+
+@dataclass
+class SharedContext:
+    """Shared context accessible via tools."""
+
+    context: dict[str, dict[str, Any]] = field(default_factory=dict)
+    node_history: list[SwarmNode] = field(default_factory=list)
+    current_task: str | None = None
+    available_nodes: list[SwarmNode] = field(default_factory=list)
+
+    def set_task(self, task: str) -> None:
+        """Set the current task."""
+        self.current_task = task
+
+    def set_available_nodes(self, nodes: list[SwarmNode]) -> None:
+        """Set list of available agents."""
+        self.available_nodes = nodes
+
+    def add_context(self, node: SwarmNode, key: str, value: Any) -> None:
+        """Add context."""
+        if node.node_id not in self.context:
+            self.context[node.node_id] = {}
+        self.context[node.node_id][key] = value
+
+    def get_relevant_context(self, target_node: SwarmNode) -> dict[str, Any]:
+        """Get context relevant to specific node."""
+        return {
+            "task": self.current_task,
+            "node_history": [node.node_id for node in self.node_history],
+            "shared_context": {k: v for k, v in self.context.items() if v},
+            "available_nodes": [node.node_id for node in self.available_nodes if node != target_node],
+        }
+
+
+@dataclass
+class SwarmState:
+    """Current state of swarm execution."""
+
+    current_node: SwarmNode
+    task: str
+    completion_status: Status = Status.PENDING
+    shared_context: SharedContext = field(default_factory=SharedContext)
+    node_history: list[SwarmNode] = field(default_factory=list)
+    message_history: list[SwarmMessage] = field(default_factory=list)
+    iteration_count: int = 0
+    start_time: float = field(default_factory=time.time)
+    last_node_sequence: list[SwarmNode] = field(default_factory=list)
+    final_result: str | None = None
+    results: dict[str, NodeResult] = field(default_factory=dict)
+    accumulated_usage: Usage = field(default_factory=lambda: Usage(inputTokens=0, outputTokens=0, totalTokens=0))
+    accumulated_metrics: Metrics = field(default_factory=lambda: Metrics(latencyMs=0))
+    execution_count: int = 0
+    execution_time: int = 0
+
+    def should_continue(self, config: SwarmConfig) -> Tuple[bool, str]:
+        """Check if the swarm should continue.
+
+        Returns: (should_continue, reason)
+        """
+        elapsed = time.time() - self.start_time
+
+        # 1. Check completion status
+        if self.completion_status != Status.EXECUTING:
+            return False, f"completion_status_changed_to_{self.completion_status}"
```

Comment:
Nit, but from the api perspective,t his seems more like:

```suggestion
            return False, f"completion_status_is_{self.completion_status}"
```

Reply:
Is there a reason that we're using underscores instead of spaces in the messages?

Reply:
no reason, will update that